src/oidcendpoint/oidc/token.py

```diff
@@ -54,21 +94,16 @@ def _access_token(self, req, **kwargs):
                 error="invalid_request", error_description="Missing code"
             )
 
-        # Session might not exist or _access_code malformed
-        try:
-            _info = _sdb[_access_code]
-        except KeyError:
-            return self.error_cls(
-                error="invalid_grant", error_description="Code is invalid"
-            )
+        _session_info = _mngr.get_session_info_by_token(_access_code)
+        code = _mngr.find_token(_session_info["session_id"], _access_code)
+        _authn_req = _session_info["client_session_info"]["authorization_request"]
 
-        _authn_req = _info["authn_req"]
-
-        # assert that the code is valid
-        if _context.sdb.is_session_revoked(_access_code):
-            return self.error_cls(
-                error="invalid_grant", error_description="Session is revoked"
-            )
+        if "state" in req:
+            # verify that state in this request is the same as the one in the
+            # authorization request
+            if _authn_req["state"] != req["state"]:
```

Comment:
What if `state` is defined in `_authn_req` but not in `req`? Then we won't throw an error.

Also what if `state` is not defined in `_authn_req` but it is in `req`? A KeyError will be raised.

We should write tests for these cases. I can do that if you want